azurerm/internal/services/mssql/mssql_server_transparent_data_encryption_resource.go

```diff
@@ -0,0 +1,213 @@
+	var encryptionProtectorProperties sql.EncryptionProtectorProperties
+
+	if v, ok := d.GetOk("key_vault_uri"); ok {
+		keyUri := v.(string)
+
+		// Set the SQL Server Key properties
+		serverKeyProperties := sql.ServerKeyProperties{
+			ServerKeyType: sql.AzureKeyVault,
+			URI:           &keyUri,
+		}
+		serverKey.ServerKeyProperties = &serverKeyProperties
+
+		// Set the encryption protector properties
+		keyDetails, err := keyVaultParser.ParseNestedItemID(keyUri)
+
+		if err != nil {
+			return fmt.Errorf("Unable to parse key uri: %q: %+v", keyUri, err)
+		}
+
+		// Make sure it's a key, if not, throw an error
+		if keyDetails.NestedItemType == "keys" {
+			keyName := keyDetails.Name
+			keyVersion := keyDetails.Version
+
+			// Extract the vault name from the keyvault base url
+			idURL, err := url.ParseRequestURI(keyDetails.KeyVaultBaseUrl)
+
+			if err != nil {
+				return fmt.Errorf("Unable to parse key vault hostname: %s", keyDetails.KeyVaultBaseUrl)
+			}
+
+			hostParts := strings.Split(idURL.Host, ".")
+			vaultName := hostParts[0]
+
+			// Create the key path as for the Encryption Protector. Format is: {vaultname}_{key}_{key_version}
+			serverKeyName = fmt.Sprintf("%s_%s_%s", vaultName, keyName, keyVersion)
+
+			// Set the encryption protector properties
+			encryptionProtectorProperties = sql.EncryptionProtectorProperties{
+				ServerKeyType: sql.AzureKeyVault,
+				ServerKeyName: &serverKeyName,
+			}
+		} else {
+			return fmt.Errorf("Key vault uri must be a reference to a key, but got: %s", keyDetails.NestedItemType)
+		}
+	} else {
+		serverKeyName = ""
+
+		// Service managed doesn't require a key name
+		encryptionProtectorProperties = sql.EncryptionProtectorProperties{
+			ServerKeyType: sql.ServiceManaged,
+			ServerKeyName: &serverKeyName,
+		}
+	}
+
+	// Only create  aserver key
+	if serverKey.ServerKeyProperties != nil {
+		// Create a key on the server
+		futureServers, err := serverKeysClient.CreateOrUpdate(ctx, resGroup, serverName, serverKeyName, serverKey)
+		if err != nil {
+			return err
+		}
+
+		if err = futureServers.WaitForCompletionRef(ctx, serverKeysClient.Client); err != nil {
+			return err
+		}
+	}
+
+	encryptionProtectorObject := sql.EncryptionProtector{
+		EncryptionProtectorProperties: &encryptionProtectorProperties,
+	}
+
+	futureEncryptionProtector, err := encryptionProtectorClient.CreateOrUpdate(ctx, resGroup, serverName, encryptionProtectorObject)
+	if err != nil {
+		return err
+	}
+
+	if err = futureEncryptionProtector.WaitForCompletionRef(ctx, encryptionProtectorClient.Client); err != nil {
+		return fmt.Errorf("waiting on create/update future for Encryption Protector on server %q (Resource Group %q): %+v", serverName, resGroup, err)
+	}
+
+	resp, err := encryptionProtectorClient.Get(ctx, resGroup, serverName)
+	if err != nil {
+		return fmt.Errorf("issuing get request for Encryption protector on server %q (Resource Group %q): %+v", serverName, resGroup, err)
+	}
+
+	d.SetId(*resp.ID)
```

Comment:
Can we make use of the generated ID functions here?  You should have enough information to use 
```
id := NewEncryptionProtectorID(serverId.Subscription, serverId.ResourceGroup, serverId.Name, name)
```

```suggestion
	d.SetId(id.ID())
```

Reply:
Updated